### ISSUE TITLE
PatchAffineShapeEstimator returns NaN for every float16 patch (moments underflow + eps=1e-10 flushes to zero)

### ISSUE THREAD

**Issue by ducha-aiki** [kornia/kornia#4123]
## Describe the bug

`PatchAffineShapeEstimator` returns NaN for every patch in `float16`, through two stacked failures in `forward` (`kornia/feature/affine_shape.py`):

1. The weighted-gradient second moments underflow: the Gaussian `weighting` peaks at ~1.1e-3, so weighted gradients are ~5e-4 and their squares ~3e-7; averaged over the patch, all three moments flush to exactly `0.0` in fp16.
2. The degenerate-patch guard that should catch this — `ellipse_shape < self.eps` with `eps = 1e-10` — never fires, because **1e-10 itself flushes to 0.0 when compared in fp16** (`0.0 < 0.0` is false). The circular-shape replacement is skipped and the final `ellipse_shape / ellipse_shape.max(...)` divides 0/0 → NaN.

The NaN then propagates through `ellipse_to_laf` into `LAFAffineShapeEstimator`'s output, so every fp16 LAF comes back NaN.

Until #4122, this was masked on CPU by `torch.inverse` raising `linalg.inv: Low precision dtypes not supported` first — the pipeline failed either way, but with an error instead of silent NaN. With the closed-form `ellipse_to_laf`, the NaN is what reaches the caller.

## Reproduction

```python
import torch
from kornia.feature.affine_shape import PatchAffineShapeEstimator

patch = torch.zeros(1, 1, 32, 32, dtype=torch.float16)
patch[:, :, 15:-15, 9:-9] = 1  # a perfectly good, high-contrast patch
print(PatchAffineShapeEstimator(32).to(torch.float16)(patch))
# tensor([[[nan, nan, nan]]], dtype=torch.float16)
```

`bfloat16` does not NaN (its fp32-like exponent range keeps both the moments and the eps alive) but still fails the `test_toy` tolerance override (`atol=1e-4`, tighter than `testing/base.py`'s bfloat16 default of 7.8e-3 — the #4081 override pattern).

## Expected behavior

Finite ellipse shapes in fp16. The moments carry the actual shape signal, so the fix is fp32-internal math (compute the second moments and the eps comparison in `float32`, cast the normalized result back) rather than just an eps bump — an eps bump alone would classify *every* fp16 patch as degenerate and return circles.

Environment: kornia at `2009933e`, torch 2.9.1, CPU (device-independent arithmetic).

**Comment by ducha-aiki** [comment]
Fixed on `main` — closing.

The fix landed inside #4122 (`1331a861`), not as a standalone PR, so this issue was never auto-closed. `PatchAffineShapeEstimator.forward` now promotes `float16`/`bfloat16` patches to `float32` and rebuilds the Gaussian weighting at `float32` before computing the moments, casting the normalized result back at the end — i.e. the fp32-internal approach this issue asked for, not an eps bump. #4131 proposed the same fix independently and was closed in favour of the version that shipped.

Verified on `main` @ `8db1499a`, torch 2.14.0, Apple silicon, with the reproduction from this issue (`patch[:, :, 15:17, 9:23] = 1`):

| device | float16 | bfloat16 | float32 reference |
|---|---|---|---|
| CPU | `[0.08832, 0.0, 1.0]` | `[0.08838, 0.0, 1.0]` | `[0.088334, 0.0, 1.0]` |
| MPS | `[0.08832, 0.0, 1.0]` | `[0.08838, 0.0, 1.0]` | `[0.088334, 0.0, 1.0]` |

Both failure modes named in the report are gone: the CPU `[nan, nan, nan]` and the MPS `[1.0, 0.0, 1.0]` circular-shape fallback (the case a finite-output check alone would have missed). A genuinely flat patch still returns `[1.0, 0.0, 1.0]` on both devices, so the degenerate branch still works.

The regression is pinned rather than merely fixed. `tests/feature/test_affine_shape_estimator.py` now carries `test_half_precision_retains_anisotropic_shape`, `test_half_precision_uses_float32_weighting`, `test_zero_patch_uses_circular_shape` and `test_degenerate_ellipse_falls_back_to_input_laf_float16`; I checked that they discriminate by temporarily disabling the fp32 promotion, which re-breaks 5 of them. The file is 50 passed / 17 skipped across float16, bfloat16, float32 and float64 on CPU.

One related item stays open and is **not** part of this issue: `TestLAFAffNetShapeEstimator::test_toy[cpu-float16]` and `[cpu-bfloat16]` are in the `testing/half_precision_xfails/` manifests from #4154. That is the AffNet path failing a hard-coded `atol=1e-4` (the #4081 override pattern — the sibling `TestLAFAffineShapeEstimator::test_toy` in the same file already uses per-dtype tolerances), not a numerical defect in the shape estimator. Tracked under #4153.

Posted on behalf of @ducha-aiki by Claude (Opus 5, 1M context).
